lock-api/src/main/java/com/palantir/lock/client/AsyncTimeLockUnlocker.java

```diff
@@ -39,59 +36,59 @@
  * There is another layer of retrying below us (at the HTTP client level) for external timelock users.
  * Also, in the event we fail to unlock (e.g. because of a connection issue), locks will eventually time-out.
  * Thus not retrying is reasonably safe (as long as we can guarantee that the lock won't otherwise be refreshed).
+ *
+ * Concurrency: We want to guarantee that a token T that is enqueued is included in some call to unlockOutstanding.
+ * If T can pass the compareAndSet, then T itself is scheduled. If T does not, that means there is some other
+ * thread that has scheduled the task, but the task has not retrieved the reference to the set of tokens to be
+ * unlocked (because it re-sets unlockIsScheduled to true first, before extracting the reference to the set).
  */
 public class AsyncTimeLockUnlocker implements AutoCloseable {
     private static final Logger log = LoggerFactory.getLogger(AsyncTimeLockUnlocker.class);
 
-    private static final Duration KICK_JOB_INTERVAL = Duration.ofSeconds(1);
-
     private final TimelockService timelockService;
     private final ScheduledExecutorService scheduledExecutorService;
 
-    private final AtomicBoolean available = new AtomicBoolean(true);
-    private final AtomicReference<Set<LockToken>> outstandingLockTokens;
+    private final AtomicBoolean unlockIsScheduled = new AtomicBoolean(false);
 
-    AsyncTimeLockUnlocker(TimelockService timelockService, ScheduledExecutorService scheduledExecutorService) {
-        this(timelockService, scheduledExecutorService, new AtomicReference<>(ImmutableSet.of()));
-    }
+    // Fairness incurs a performance penalty but we do not want to starve the actual unlocking process.
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock(true);
 
-    @VisibleForTesting
-    AsyncTimeLockUnlocker(TimelockService timelockService,
-            ScheduledExecutorService scheduledExecutorService,
-            AtomicReference<Set<LockToken>> outstandingLockTokens) {
+    private Set<LockToken> outstandingLockTokens = Sets.newConcurrentHashSet();
+
+    AsyncTimeLockUnlocker(TimelockService timelockService, ScheduledExecutorService scheduledExecutorService) {
         this.timelockService = timelockService;
         this.scheduledExecutorService = scheduledExecutorService;
-        this.outstandingLockTokens = outstandingLockTokens;
-        schedulePeriodicKickJob();
     }
 
     /**
      * Adds all provided lock tokens to a queue to eventually be scheduled for unlocking.
      * Locks in the queue are unlocked asynchronously, and users must not depend on these locks being unlocked /
-     * available for other users immediately.
+     * unlockIsScheduled for other users immediately.
```

Comment:
I think this was an over-zealous find-replace :)

Reply:
Yep, I'm surprised this hit for `available` which isn't such a common word! Fixed.